util/patricia_trie/src/lib.rs

```diff
@@ -49,155 +49,159 @@ mod lookup;
 			TrieError::IncompleteDatabase(_) => "Incomplete database",
-			TrieError::DecoderError(ref e) => e.description(),
+			TrieError::DecoderError(_, ref err_string) => err_string,
 		}
 	}
 }
 
-impl From<rlp::DecoderError> for Box<TrieError> {
-	fn from(e: rlp::DecoderError) -> Self { Box::new(TrieError::DecoderError(e)) }
-}
-
-/// Trie result type. Boxed to avoid copying around extra space for `H256`s on successful queries.
-pub type Result<T> = ::std::result::Result<T, Box<TrieError>>;
+/// Trie result type. Boxed to avoid copying around extra space for the `Hasher`s `Out`s on successful queries.
+pub type Result<T, U> = ::std::result::Result<T, Box<TrieError<U>>>;
 
-/// Trie-Item type.
-pub type TrieItem<'a> = Result<(Vec<u8>, DBValue)>;
+/// Trie-Item type used for iterators over trie data.
+pub type TrieItem<'a, U> = Result<(Vec<u8>, DBValue), U>;
 
 /// Description of what kind of query will be made to the trie.
 ///
 /// This is implemented for any &mut recorder (where the query will return
 /// a DBValue), any function taking raw bytes (where no recording will be made),
 /// or any tuple of (&mut Recorder, FnOnce(&[u8]))
-pub trait Query {
+pub trait Query<H: Hasher> {
 	/// Output item.
 	type Item;
 
 	/// Decode a byte-slice into the desired item.
-	fn decode(self, &[u8]) -> Self::Item;
+	fn decode(self, data: &[u8]) -> Self::Item;
 
 	/// Record that a node has been passed through.
-	fn record(&mut self, &H256, &[u8], u32) { }
+	#[allow(unused_variables)]
+	fn record(&mut self, hash: &H::Out, data: &[u8], depth: u32) {}
 }
 
-impl<'a> Query for &'a mut Recorder {
+impl<'a, H: Hasher> Query<H> for &'a mut Recorder<H> {
 	type Item = DBValue;
-
 	fn decode(self, value: &[u8]) -> DBValue { DBValue::from_slice(value) }
-	fn record(&mut self, hash: &H256, data: &[u8], depth: u32) {
+	fn record(&mut self, hash: &H::Out, data: &[u8], depth: u32) {
 		(&mut **self).record(hash, data, depth);
 	}
 }
 
-impl<F, T> Query for F where F: for<'a> FnOnce(&'a [u8]) -> T {
+impl<F, T, H: Hasher> Query<H> for F where F: for<'a> FnOnce(&'a [u8]) -> T {
 	type Item = T;
-
 	fn decode(self, value: &[u8]) -> T { (self)(value) }
 }
 
-impl<'a, F, T> Query for (&'a mut Recorder, F) where F: FnOnce(&[u8]) -> T {
+impl<'a, F, T, H: Hasher> Query<H> for (&'a mut Recorder<H>, F) where F: FnOnce(&[u8]) -> T {
 	type Item = T;
-
 	fn decode(self, value: &[u8]) -> T { (self.1)(value) }
-	fn record(&mut self, hash: &H256, data: &[u8], depth: u32) {
+	fn record(&mut self, hash: &H::Out, data: &[u8], depth: u32) {
 		self.0.record(hash, data, depth)
 	}
 }
 
 /// A key-value datastore implemented as a database-backed modified Merkle tree.
 pub trait Trie {
+	type H: Hasher;
 	/// Return the root of the trie.
-	fn root(&self) -> &H256;
+	fn root(&self) -> &<Self::H as Hasher>::Out;
 
 	/// Is the trie empty?
-	fn is_empty(&self) -> bool { *self.root() == KECCAK_NULL_RLP }
+	fn is_empty(&self) -> bool { *self.root() == Self::H::HASHED_NULL_RLP }
```

Comment:
specifically, this should draw from the node codec, because non-rlp-encoded tries will have non-rlp-encoded null nodes. so the hash will be different. this is an architectural issue:

`trait NodeCodec<H: Hasher>` should have the `const HASH_NULL_NODE: H::Out`. makes implementation of the trait in a generic context impossible w/o const functions but for now we can limit scope a bit.